src/terminal.ts

```diff
@@ -47,6 +58,56 @@ export abstract class Terminal implements ITerminal {
     this._checkType('uid', opt.uid ? opt.uid : null, 'number');
     this._checkType('gid', opt.gid ? opt.gid : null, 'number');
     this._checkType('encoding', opt.encoding ? opt.encoding : null, 'string');
+
+    // setup flow control handling
+    if (opt.handleFlowControl) {
+      this._flowPause = opt.flowPause || FLOW_PAUSE;
+      this._flowResume = opt.flowResume || FLOW_RESUME;
+      this.enableFlowHandling();
+    }
+  }
+
+  /**
+   * Enable automatic flow control handling.
+   * Instead of relying on XON/XOFF directly we pause/resume
+   * the underlying socket. Main advantage is that flow control will
+   * still work even when XON/XOFF is not available. As a downside it
+   * does not act immediately, instead the OS pty buffer and node's socket buffer
+   * fill up first before the slave program gets paused.
+   * Note that `flowPause` and `flowResume` must be in line with the
+   * messages sent by the terminal emulator to indicate PAUSE/RESUME.
+   * @param flowPause String to pause the pty. Defaults to XOFF (x13).
+   * @param flowResume String to resume the pty. Defaults to XON (x11).
+   */
+  public enableFlowHandling(flowPause: string = this._flowPause, flowResume: string = this._flowResume): void {
+    if (this._realWrite) {
+      return;
+    }
+    this._realWrite = this.write;
+    this.write = (data: string) => {
```

Comment:
Maybe write should be a `(data: string) => void` property on terminal.ts and it can be assigned by unix/windowsTerminal? Feels so dodgy monkey patching class methods.

Reply:
Sounds good.